### PR TITLE
API: Add ``lib.array_utils`` namespace

### PR DESCRIPTION
**Pull request by mtsokol**
Relevant issues https://github.com/numpy/numpy/issues/24507 and https://github.com/numpy/numpy/issues/24166

Hi @rgommers @ngoldbaum,

This PR adds `lib.array_utils` module as as public namespace which creates only local namespace.
Initially it hosts three functions:
- `byte_bounds` (moved from `np.lib.utils`)
- `normalize_axis_tuple`
- `normalize_axis_index`

### CODE REVIEW

File: numpy/typing/tests/data/fail/lib_utils.pyi [hunk 1]
```diff
@@ -1,5 +1,3 @@
 import numpy as np
 
-np.byte_bounds(1)  # E: incompatible type
```
Comment (rgommers) [inline numpy/typing/tests/data/fail/lib_utils.pyi:3]: Shouldn't these typing tests stay, just under the new name?

**Review by rgommers — COMMENTED** [review]
Thanks @mtsokol. This looks pretty good to me. One idea to consider: would it make sense to move what is now in `array_utils.py` to `_array_utils_impl.py` and make the content of a new  `array_utils.py` only this:
```python
from ._array_utils_impl import (
    byte_bounds,
    normalize_axis_index,
    normalize_axis_tuple,
)
```
That would make it pretty much bullet-proof again more accidental leaking in of stray objects.

**Comment by rgommers** [comment]
Also, as we just discussed, these are the first time we expose `normalize_axis_index` and `normalize_axis_tuple` as public API, so they should show up in the html docs after this PR.

File: numpy/typing/tests/data/fail/lib_utils.pyi (hunk 1 above)
Comment (mtsokol) [reply numpy/typing/tests/data/fail/lib_utils.pyi:3]: Sure! Now I import it from the new place.

**Comment by mtsokol** [comment]
> Also, as we just discussed, these are the first time we expose `normalize_axis_index` and `normalize_axis_tuple` as public API, so they should show up in the html docs after this PR.

@rgommers Sure, I added these two functions to `routines.other.rst` file.

**Comment by rgommers** [comment]
Two questions:
- Can you verify that the removal of `np.byte_bounds` is okay with `joblib` now? 
- Since this is the first new namespace, it may be useful to ping the mailing list and point them at gh-24166 regarding the `array_utils` name. Can you do that?

**Comment by mtsokol** [comment]
>     * Can you verify that the removal of `np.byte_bounds` is okay with `joblib` now?

It looks like `joblib` uses `np.byte_bounds` in two places and one `np.NaN` that has already been removed. I will created a PR to address it: https://github.com/joblib/joblib/pull/1501
 
>     * Since this is the first new namespace, it may be useful to ping the mailing list and point them at [DISCUSS: Namespace for array creation/inspection utils? #24166](https://github.com/numpy/numpy/issues/24166) regarding the `array_utils` name. Can you do that?

Sure! I sent a message to the mailing list.

**Comment by ngoldbaum** [comment]
Given this adds a new public name to the API let's let this sit for a few more days to gather comments.

File: numpy/lib/array_utils.pyi
Comment (BvB93) [inline numpy/lib/array_utils.pyi:1]: Minor nit: an `__all__: list[str]` annotation is missing.

File: numpy/lib/_array_utils_impl.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+from numpy.core import asarray
+from numpy.core.numeric import normalize_axis_tuple, normalize_axis_index
+
+
+__all__ = ["byte_bounds", "normalize_axis_tuple", "normalize_axis_index"]
+
+
+def byte_bounds(a):
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply numpy/lib/_array_utils_impl.py:9]: Sure! I added `set_module`.

[the remark this answers is not in the document]
Comment (rgommers) [reply numpy/lib/_array_utils_impl.py:9]: It looks like the two `normalize_*` functions also need that tweak to their `__module__` attribute.

[the remark this answers is not in the document]
Comment (mtsokol) [reply numpy/lib/_array_utils_impl.py:9]: Sure! I updated their `__module__` attributes.

**Review by rgommers — APPROVED** [review]
LGTM now. Everyone's comments have been addressed, and the [email to the mailing list](https://mail.python.org/archives/list/numpy-discussion@python.org/thread/BUCQLPJAEUMGMNZTKNHD4PHYSDKUFEBC/) didn't get replies. So I think we're good here. I suggest we merge this in a day or two unless there are new comments.

**Comment by ngoldbaum** [comment]
> I suggest we merge this in a day or two unless there are new comments.

In it goes. Thanks @mtsokol.

**Comment by seberg** [comment]
FWIW, the versionadded 1.13.0 are now deleted because it is uninterestingly ancient history anyway ;).

**Pull request merged**